### PR TITLE
Add a config target for the use of ICC on Windows x86_64

### LINKED ISSUE #1150: add ICC support on Windows [post 1.1.0]
Hi, Everyone,

In short, Intel provide [C, C++ compiler on Windows](https://software.intel.com/en-us/intel-parallel-studio-xe), which has [some advantages](https://software.intel.com/en-us/c-compilers/ipsxe) compared to MSVC. Would be nice, if OpenSSL add support for it.

Since ICC on Windows pretends to be MSVC (compare [ICC](https://software.intel.com/en-us/node/512855) and [MSVC](https://msdn.microsoft.com/en-us/library/9s7c9wdw.aspx) options), required changes are minimal. In file `Configurations/10-main.conf`:
- add separate section `"VC-WIN64A-ICC"`, which is a duplicate of `"VC-WIN64A"` toolchain and behaves the same,
- in section `"VC-WIN64A-ICC"` override variables:

```
cc               => "cl",
```

inherit from `"VC-common"`, with

```
cc               => "icl",
```

and

```
cflags           => ... " /O2";
```

inherit from `"VC-noCE-common"`, with

```
cflags           => ... " /O3";
```

Because of OpenSSL build error using ICC:

```
#31000: corrupt PDB file ossl_static\vc140.pdb; delete and rebuild; if problem persists, delete and try /Z7 instead
```

also needed to replace in file `Configurations/10-main.conf` strings:

```
/Fdossl_static
/Fddso
/Fdapp
```

with

```
/Fdossl_static.pdb
/Fddso.pdb
/Fdapp.pdb
```

It fix this error while has no visible regressions for MSVC builds (tested for **<Debug,Release>+<shared,static>** configurations using MSVC 2015 only).

Regards,
Alexander

### PR DESCRIPTION
**Pull request by levitte**
Note that ICC is currently not treating the /Fd value quite like MSVC,
but that's quite all right, all that's to get them working the same
way is to make the extension explicit.

Fixes #1150

-----

Note that this follows the numbering scheme from #5043, and this config target is placed in the "unknown" category until further notice.

### CODE REVIEW

**Comment by richsalz** [comment]
Until someone steps up to take it on, I think the default category should be community.  I see unknown as a holding place for things we already have that we have not evaluated.

**Review by richsalz — APPROVED** [review]

**Comment by levitte** [comment]
I think the debate is already taking place elsewhere, I see no reason why it should be held here as well.

**Comment by levitte** [comment]
Anyway, I'll wait until @sav-ix has had a chance to test it before merging

**Comment by sav-ix** [comment]
In case that matters, since 1.0.2-dev (possible earlier) till recent 1.1.1-dev, there were **any** OpenSSL build issues using ICC on Windows, which weren't duplicated by builds using MSVC. Thus OpenSSL build issues using ICC on Windows, which may arise in the future, are likely relate to the ICC itself and would be out of scope of OpenSSL.
Though if @openssl/committers would obtain ICC license for Open Source development on Windows, it will completely solve this question.

**Comment by dot-asm** [comment]
I suppose you rather meant "there were *no* build issues".

> if @openssl/committers would obtain ICC license for Open Source development on Windows, it will completely solve this question.

First of all, it's a big **if**, and secondly even if it was an option, it wouldn't actually "*completely* solve this question." We can't do *everything* and have to compromise. In other words additional toolchain has to provide some *compelling* advantage. And specifically in OpenSSL context. Does it? If so, by how much? We also have to keep in mind how wide is user base. For example if we are talking about additional Windows toolchains, then maybe clang would be more appropriate...

**Comment by dot-asm** [comment]
Besides, it's not actually exclusively about build. In the context support is rather about responding to problem reports with toolchain itself. just consider -arch:AVX2 case...

**Comment by levitte** [comment]
> I would argue that community-supported config files should be actually contributed. Because otherwise it's unclear who supports what. I mean let's say you, Richard, do this. After a while an Windows ICC user shows up with a problem. What would be his/her expectations? That Richard will solve the problem. But how would you do that? Without access to the toolchain?

Right now, I have it in the "unknown" category, following @t-j-h's idea that this is the category for config targets that don't have any supporter (sponsor?) assigned to it.
If someone, say @sav-ix, takes it upon them to run (semi)regular tests, and submits PRs when something needs fixing, then we could re-classify this config target as "community supported".

But *either way*, there's absolutely nothing stopping anyone from making a PR to add or fix as per someone else's issue, and asking that person to either confirm that it works or to report what goes wrong.  After all, this shouldn't even be strange to anyone, it's been done before.

**Comment by levitte** [comment]
(remember back in 1999, when I spent about a year sending in fixes for VMS before I became a committer?  That's community support...  and also, let's remember  that even if we are committers or OMC members, we are also part of the community)

**Comment by sav-ix** [comment]
> I suppose you rather meant "there were no build issues".

yes, thanks.

> it's a big if

Recently it wasn't a big **if**; we can clarify, how things are now.

> We can't do everything and have to compromise.

ICC on Windows imitates MSVC (including use the same Windows SDK, as MSVC do), hence the connectedness of its errors with MSVC ones, when they arise. So until OpenSSL supports MSVC, expected resource consumption of Windows ICC support should be minimal.

> In other words additional toolchain has to provide some compelling advantage

Does this mean, that the current set of OpenSSL supported compilers was made taking into account such requirements? If so, is there a rating of OpenSSL supported compilers, depending of provided advantages?
And in essence, ICC provides some advantages compared, say, MSVC (profiling tools, better language support, MPI, MKL, DAAL libraries, [etc.](https://software.intel.com/en-us/c-compilers/ipsxe/details)). But they unlikely has a significant effect on OpenSSL builds.

> We also have to keep in mind how wide is user base.

Welcome to [Intel forums](https://software.intel.com/en-us/forums/intel-software-development-products).

> if we are talking about additional Windows toolchains, then maybe clang would be more appropriate

Clang is amazing regardless of the platform. But whether it is necessary to oppose them to each other and choose one.

> just consider -arch:AVX2 case...

Know nothing special with [it](https://github.com/openssl/openssl/issues/3803). It took 11 posts (could meet the 3) to figure out, that issue is out of scope. Further discussion was at will (mean was not compulsory).

**Comment by dot-asm** [comment]
Thought experiment. I think you, Richard, should add gcc support for VMS ;-) If not, why [and how is it different from this case]?

**Comment by dot-asm** [comment]
> Clang is amazing regardless of the platform. But whether it is necessary to oppose them to each other and choose one.

*All* questions can be resolved if you, @sav-ix, explicitly say "I undertake the task of periodically test this for conceivable eternity." But you don't say that, you keep implying that project members should get the license and undertake the task. Unfortunately we are limited in resources [not to mention that none of us use even Windows as development platform on regular basis], and as long as we are, we will have to oppose things to each other. It's life...

> And in essence, ICC provides some advantages compared, say, MSVC (profiling tools, better language support, MPI, MKL, DAAL libraries, etc.).

And what prevents you from compiling OpenSSL with MSVC and link it with your application code that utilizes MPI, MKL, DAAL, etc?

> But they unlikely has a significant effect on OpenSSL builds.

Exactly my point. It doesn't provide any compelling advantage.

> > just consider -arch:AVX2 case...

> Know nothing special with it.

You have to imagine how would it look like if I had no option to test it, or if problem couldn't be reproduced with VS2017 that I have...

**Comment by levitte** [comment]
Simple, @dot-asm, there is no gcc for VMS, hasn't been for ages.  But if *you* could instruct me what was needed, I could certainly try and help creating a config target.  That's a moot point, though, 'cause you know this stuff just as well as I, so...

**Comment by sav-ix** [comment]
> Right now, I have it in the "unknown" category

Assume it would suite to any category, which does not impose obligations on its official support. At least until Intel would provide appropriate license for OpenSSL Project.

> If someone, say @sav-ix, takes it upon them to run (semi)regular tests, and submits PRs when something needs fixing, then we could re-classify this config target as "community supported".

For myself I can say, that run OpenbSSL builds, including with Windows ICC, on average once a month. But for now I'm ready only to report issues and provide testing for proposed patches, but not publish PR. And if this is a mandatory requirement, then perhaps it make sense to put this PR on pause until such a person appears.

**Comment by dot-asm** [comment]
> there is no gcc for VMS

??? But there are VMS support files in gcc source code tree...

> But if you could instruct me what was needed,

https://gcc.gnu.org/onlinedocs/gcc-3.0/gcc_4.html#SEC63

But let's not forget that this is "thought experiment" ;-)

**Comment by richsalz** [comment]
As @dot-asm said, it's not about getting the license to use the compiler.  it's about the team making a real commitment to regularly test and build against that compiler, and nobody is signing up to do that.  The only practical alternative, then, is for you to make pull requests to add the configuration and fix bugs as you find them:  by definition (see https://www.openssl.org/policies/platformpolicy.html) this is a community supported platform.

If you can't agree to take on the burden, then it won't go into the official releases.

**Comment by levitte** [comment]
Hey @sav-ix, I did find a stoopid fault in our `build.info` files that must have affected your build badly, like you reported in #1150.  For certain parts of the build, we depended on the config target name starting with `VC-`, when we really should look for `-WIN(32|64)`.  I've added a commit that fixes this, so if you'd please try again?

**Comment by dot-asm** [comment]
> > Right now, I have it in the "unknown" category

I would like to argue against *adding* things to "unknown" category. Primary use for that should rather be to phase *out* old and obsolete things. If we are to learn anything from the current state of affairs, it's that it's extremely hard to remove things. So why adding shouldn't have some kind of barrier?

**Comment by sav-ix** [comment]
> @dot-asm Unfortunately we are limited in resources
> @richsalz If you can't agree to take on the burden, then it won't go into the official releases.

That's understandable. Additional arguments to put this PR on pause.

> @dot-asm And what prevents you from compiling OpenSSL with MSVC and link it with your application code

A wish to use binary files of different libraries, built using the same compiler with the same set of flags in order to decrease likelihood of possible runtime errors.

> @levitte I've added a commit that fixes this, so if you'd please try again?

Sure. Not more than assumption, such (OpenSSL build system) errors could be revealed using MSVC too. Mean set `cc               => "cl",` variable in `ICC-WIN64` toolchain, an run build using MSVC, as if it's ICC on Windows.

**Comment by levitte** [comment]
> by definition (see https://www.openssl.org/policies/platformpolicy.html) this is a community supported platform.

... or "unknown"

**Comment by dot-asm** [comment]
> A wish to use binary files of different libraries, built using the same compiler with the same set of flags in order to decrease likelihood of possible runtime errors.

Then why don't you call Microsoft and ask for kernel32.dll compiled with ICC? :-) What I'm saying is that it's perfectly possible to use libraries compiled with different compiler flags (likely modulo /M\* flag), and even with different compilers and versions. If it wasn't possible, very little would actually work... You should have more faith in ABI compliance :-)

**Comment by dot-asm** [comment]
BTW, what prevents user from calling `perl Configure VC-WIN64A CC=icl` as it is *now*?

**Comment by sav-ix** [comment]
Applied PR# https://github.com/openssl/openssl/pull/5058/commits/0d5bc8b9a293baba842b9d5724cf21e405a572b3 , and for builds using ICC on Windows with Shared layout error not reproduced.
Now both Shared and Static builds fails with the same error:
```
Can't Open ./tools/c_rehash.pl at .\util\copy.pl line 62.
```
during installation.

**Comment by sav-ix** [comment]
> BTW, what prevents user from calling perl Configure VC-WIN64A CC=icl as it is now?

This it work fine, though other compilers doesn't require to define compiler name in build commands. But related [patch](https://github.com/openssl/openssl/issues/1150#issuecomment-356997420) also provided additional changes, and it wasn't expected, that some of them would land in MSVC toolchain directly.

**Comment by sav-ix** [comment]
> You should have more faith in ABI compliance

If it's possible to decrease chance of spending time in disassembler because of compatibility issues in backends of different compilers, it make sense to use it.

**Comment by dot-asm** [comment]
>> BTW, what prevents user from calling perl Configure VC-WIN64A CC=icl as it is now?

> This it work fine, though other compilers doesn't require to define compiler name in build commands.

`perl Configure linux-x86_64 CC=icc` or `env CC=icc config`. If I'm to test specific version I customarily do for example `env CC=gcc-4.8 config`...

> But related patch also provided additional changes, and it wasn't expected, that some of them would land in MSVC toolchain directly.

If those /Fd can be *confirmed* *not* to ruin things for MSVC, e.g. by producing foo.pdb.pdb, then it's not really a problem to do apply *that* modification. Keyword is "**confirmed**". As for /O3 it's *your* preference, and it can be handled as any other preference, by passing it as additional argument at configuration time. It will appear later and will override default.

**Comment by sav-ix** [comment]
> If those /Fd can be confirmed not to ruin things for MSVC

But there could be workarounds, which not supported by MSVC (e.g. ` -Qoption,cpp,--unicode_source_kind,UTF-8` flag, as a workaround for MSVC [-utf-8](https://msdn.microsoft.com/en-us/library/mt708821.aspx) flag). They obviously can't be added to MSVC toolchain, since produce noise for MSVC builds.

> As for /O3 it's your preference, and it can be handled as any other preference

Most OpenSSL compiler toolchains use maximum available optimization level by default. ICC shouldn't be an exception.

> It will appear later and will override default.

Yes, this works too. But as a side effect it would result in noise in stderr like `-O2 flag was overridden with -O3 ...` per each compilation unit.
Surely, such warnings could be silenced by appropriate flag. But why all of this, if it's possible to add a separate toolchain for Windows ICC and don't overload build commands with additional flags?

**Comment by levitte** [comment]
> ```
> Can't Open ./tools/c_rehash.pl at .\util\copy.pl line 62.
> ```

I just committed a fix, a couple spots that were still relying on the config target name starting with `VC-`

**Comment by levitte** [comment]
> If those /Fd can be *confirmed* *not* to ruin things for MSVC, e.g. by producing foo.pdb.pdb, then it's not really a problem to do apply *that* modification. Keyword is "**confirmed**".

According to [docs (this is for VC 2008)](https://msdn.microsoft.com/en-us/library/9wst99a9(v=vs.90).aspx), it should be fine.  Of course, docs isn't always reality, so please do check if you have an oldie around...

Quoted from those does:

> If you specify a file name that does not include an extension, the compiler uses .pdb as the extension.

File: Configurations/10-main.conf
Comment (levitte) [reply Configurations/10-main.conf]: Depends.  If, like the MSVC docs will tell you, an explicit extension doesn't make a difference for MSVC but may make a difference for other compatible compilers, I'd say this should go into the "base class"

**Comment by sav-ix** [comment]
Applied PRs:
5a7d0d8eabf909282d56ce185b3a2da4a01ede03
0d5bc8b9a293baba842b9d5724cf21e405a572b3
c77b1b6891e079ed9027383874b0db10b25237a2
to OpenSSL-1.1.1-dev (a41a6120cdcb7e883481bc1bed55e7157c9255c4), but
```
Can't Open ./tools/c_rehash.pl at .\util\copy.pl line 62.
```
error still reproduce for all builds using Windows ICC.

**Comment by dot-asm** [comment]
Closing in favour of #5574.

**Pull request closed without merging**
